test/test_dynamic_shape_models.py

```diff
@@ -0,0 +1,90 @@
+import sys
+import numpy as np
+import unittest
+import torch
+import torch_xla.core.xla_model as xm
+import torch_xla.debug.metrics as met
+
+xla_dev = xm.xla_device()
+
+
+class Feedforward(torch.nn.Module):
+
+  def __init__(self, input_size, hidden_size):
+    super().__init__()
+    self.input_size = input_size
+    self.hidden_size = hidden_size
+    self.fc1 = torch.nn.Linear(self.input_size, self.hidden_size)
+    self.fc1.weight.data.fill_(0.01)
+    self.fc1.bias.data.fill_(0.01)
+    self.relu = torch.nn.ReLU()
+    self.fc2 = torch.nn.Linear(self.hidden_size, 1)
+    self.fc2.weight.data.fill_(0.01)
+    self.fc2.bias.data.fill_(0.01)
+    self.sigmoid = torch.nn.Sigmoid()
+
+  def forward(self, x):
+    hidden = self.fc1(x)
+    relu = self.relu(hidden)
+    output = self.fc2(relu)
+    output = self.sigmoid(output)
+    return output
+
+
+class TestDynamicShapeModels(unittest.TestCase):
+
+  def test_forward_pass_dynamic_input_correctness(self):
+    losses = []
+    for dev in [torch.device('cpu'), xla_dev]:
+      num_features = 2
+      num_test_samples = 5
+      x_test, y_test = self.create_dynamic_test_data(num_test_samples,
+                                                     num_features, dev)
+
+      model = Feedforward(num_features, hidden_size=10).to(dev)
+      criterion = torch.nn.BCELoss()
+
+      model.eval()
+      with torch.no_grad():
+        y_pred = model(x_test)
+        before_train = criterion(y_pred.squeeze(), y_test)
+        xm.mark_step()
+        losses.append(before_train.item())
+
+    np.testing.assert_allclose(losses[0], losses[1], rtol=1e-2, atol=1e-2)
+
+  def test_forward_pass_dynamic_input_compile_once(self):
+    met.clear_metrics()
+    losses = []
+    for _ in range(10):
+      num_features = 2
+      num_test_samples = 5
+      x_test, y_test = self.create_dynamic_test_data(num_test_samples,
+                                                     num_features, xla_dev)
+
+      model = Feedforward(num_features, hidden_size=10).to(xla_dev)
+      criterion = torch.nn.BCELoss()
+
+      model.eval()
+      with torch.no_grad():
+        y_pred = model(x_test)
+        before_train = criterion(y_pred.squeeze(), y_test)
+        xm.mark_step()
+    np.testing.assert_equal(met.metric_data('CompileTime')[0], 3)
```

Comment:
why is it 3 here?

Reply:
Does the "CompileTime" here refer to compiling IR graph to HLO graph, or compiling HLO to LLO/executable?

The [IR dump](https://paste.googleplex.com/6582669499432960) shows 3 graphs: 2 for `before_train = criterion(y_pred.squeeze(), y_test)` and 1 for `xm.mark_step()`. The "CompileTime" doesn't grow linearly with the number of iteration. Does `3` match your expectation?

Reply:
hmm, so the second graph
```
[ScheduleSyncTensorsGraph]
TensorsGraphInfo:
  __bool__ (/home/ptxla/.local/lib/python3.8/site-packages/torch/__init__.py:212)
  binary_cross_entropy (/home/ptxla/.local/lib/python3.8/site-packages/torch/nn/functional.py:3087)
  forward (/home/ptxla/.local/lib/python3.8/site-packages/torch/nn/modules/loss.py:619)
  _call_impl (/home/ptxla/.local/lib/python3.8/site-packages/torch/nn/modules/module.py:1480)
  test_forward_pass_dynamic_input_compile_once (pytorch/xla/test/test_dynamic_shape_models.py:71)
  _callTestMethod (/usr/local/lib/python3.8/unittest/case.py:633)
  run (/usr/local/lib/python3.8/unittest/case.py:676)
  __call__ (/usr/local/lib/python3.8/unittest/case.py:736)
  run (/usr/local/lib/python3.8/unittest/suite.py:122)
  __call__ (/usr/local/lib/python3.8/unittest/suite.py:84)
  run (/usr/local/lib/python3.8/unittest/suite.py:122)
  __call__ (/usr/local/lib/python3.8/unittest/suite.py:84)
  run (/usr/local/lib/python3.8/unittest/runner.py:176)
  runTests (/usr/local/lib/python3.8/unittest/main.py:271)
  __init__ (/usr/local/lib/python3.8/unittest/main.py:101)
  <module> (pytorch/xla/test/test_dynamic_shape_models.py:93)
 
Hashes: (5c2a92a233f40275064b7ca64d2c16ba)
 
## BEGIN_GRAPH
IR {
  %0 = f32[1]{0} xla::device_data(), location=convert@module.py:1128, device=TPU:0
  %1 = f32[1,10]{1,0} xla::device_data(), location=convert@module.py:1128, device=TPU:0
  %2 = f32[10,1]{0,1} aten::permute(%1), location=forward@linear.py:114, dims=(1, 0)
  %3 = f32[10]{0} xla::device_data(), location=convert@module.py:1128, device=TPU:0
  %4 = f32[10,2]{0,1} xla::device_data(), location=convert@module.py:1128, device=TPU:0
  %5 = f32[2,10]{1,0} aten::permute(%4), location=forward@linear.py:114, dims=(1, 0)
  %6 = f32[5,2]{0,1} xla::device_data(), location=create_dynamic_test_data@test_dynamic_shape_models.py:85, device=TPU:0
  %7 = s32[5,2]{0,1} xla::cast(%6), location=create_dynamic_test_data@test_dynamic_shape_models.py:86, type=s32, dtype=Int, stype=Float
  %8 = (s32[<=10,2]{1,0}, s32[]) aten::nonzero(%7), num_outputs=2, location=create_dynamic_test_data@test_dynamic_shape_models.py:86
  %9 = f32[<=10,2]{1,0} xla::cast(%8.0), location=create_dynamic_test_data@test_dynamic_shape_models.py:86, type=f32, dtype=Float, stype=Int
  %10 = f32[<=10,10]{1,0} aten::addmm(%9, %5, %3), location=forward@linear.py:114
  %11 = f32[<=10,10]{1,0} aten::relu(%10), location=relu@functional.py:1457
  %12 = f32[<=10,1]{1,0} aten::addmm(%11, %2, %0), location=forward@linear.py:114
  %13 = f32[<=10,1]{1,0} aten::sigmoid(%12), location=forward@activation.py:294
  %14 = f32[<=10]{0} aten::view(%13), location=binary_cross_entropy@functional.py:3087, output_size=(10)
  %15 = s32[] aten::size(%14), ROOT=0
}
```

is a bit concerning, it seems like we materalize the `size` by a `bool` operator somewhere. I would like to understand where does that happen in a follow up pr.